File: backend/app/core/languages.py
"""Languages the app is offered in.

One registry, used by the API, the translation cache and (mirrored) by the
client. Adding a language is adding a row here — nothing else in the system
needs to know the list.

English is the SOURCE language, not merely one of the options: the agents
write their analysis in English and everything else is a translation of it.
That is deliberate. The models reason about finance most reliably in
English, the terminology is native to it, and translating out of English is
cleaner than translating into it.

Choosing a language here is a product decision, not a legal one. Shipping
the app in German does not make it lawful to offer investment analysis in
Germany — distribution has to be restricted per country in App Store Connect
and the Play Console, separately and deliberately.
"""
from dataclasses import dataclass
from typing import Dict, List


@dataclass(frozen=True)
class Language:
    code: str
    #: Name in the language itself — what a speaker recognises in a picker.
    native_name: str
    english_name: str
    rtl: bool = False


#: Ordered: English first as the source, then by expected audience size.
LANGUAGES: List[Language] = [
    Language("en", "English", "English"),
    Language("he", "עברית", "Hebrew", rtl=True),
    Language("de", "Deutsch", "German"),
    Language("es", "Español", "Spanish"),
    Language("pt-BR", "Português (Brasil)", "Portuguese (Brazil)"),
    Language("fr", "Français", "French"),
    # The app is already built for right-to-left because of Hebrew, so Arabic
    # costs almost nothing structurally — for most apps it is a large rewrite.
    Language("ar", "العربية", "Arabic", rtl=True),
    Language("it", "Italiano", "Italian"),
    Language("ko", "한국어", "Korean"),
    Language("ja", "日本語", "Japanese"),
]

BY_CODE: Dict[str, Language] = {lang.code: lang for lang in LANGUAGES}

SOURCE_LANGUAGE = "en"
DEFAULT_LANGUAGE = "en"

RTL_CODES = {lang.code for lang in LANGUAGES if lang.rtl}


def is_supported(code: str | None) -> bool:
    return bool(code) and code in BY_CODE


def normalize(code: str | None) -> str:
    """Best supported match for a device or profile language tag.

    Devices report tags like "he-IL", "pt-BR", "en-US". An exact match wins;
    otherwise the base subtag is tried, so "de-AT" lands on German rather
    than silently falling back to English. Anything unknown becomes the
    default — never an error, because a language we do not have is a normal
    thing for a store app to meet.
    """
    if not code:
        return DEFAULT_LANGUAGE
    tag = code.strip()
    if tag in BY_CODE:
        return tag
    # Case-insensitive exact match ("PT-br").
    for known in BY_CODE:
        if known.lower() == tag.lower():
            return known
    base = tag.split("-")[0].split("_")[0].lower()
    if base in BY_CODE:
        return base
    # A regional variant of a language we carry only regionally: pt-PT → pt-BR
    # is a better answer than English for a Portuguese speaker.
    for known in BY_CODE:
        if known.split("-")[0].lower() == base:
            return known
    return DEFAULT_LANGUAGE

File: frontend/src/i18n/languages.ts
/**
 * Languages, and which one this reader gets.
 *
 * Mirrors backend/app/core/languages.py. The server is authoritative — the
 * picker is filled from GET /auth/languages — but this list exists so the
 * app can pick a language on the very first paint, before any request has
 * returned. Keep the two in step when adding a language.
 */

export type Language = {
  code: string;
  nativeName: string;
  englishName: string;
  rtl?: boolean;
};

/** English first: it is the source language every analysis is written in. */
export const LANGUAGES: Language[] = [
  { code: "en", nativeName: "English", englishName: "English" },
  { code: "he", nativeName: "עברית", englishName: "Hebrew", rtl: true },
  { code: "de", nativeName: "Deutsch", englishName: "German" },
  { code: "es", nativeName: "Español", englishName: "Spanish" },
  { code: "pt-BR", nativeName: "Português (Brasil)", englishName: "Portuguese (Brazil)" },
  { code: "fr", nativeName: "Français", englishName: "French" },
  { code: "ar", nativeName: "العربية", englishName: "Arabic", rtl: true },
  { code: "it", nativeName: "Italiano", englishName: "Italian" },
  { code: "ko", nativeName: "한국어", englishName: "Korean" },
  { code: "ja", nativeName: "日本語", englishName: "Japanese" },
];

export const DEFAULT_LANGUAGE = "en";

const BY_CODE = new Map(LANGUAGES.map((l) => [l.code, l]));

export const isRTL = (code: string): boolean => Boolean(BY_CODE.get(code)?.rtl);

/**
 * Best supported match for a device tag.
 *
 * Devices report "he-IL", "de-AT", "pt-PT". An exact match wins, then the
 * base subtag, then any regional variant we carry — so a Portuguese speaker
 * in Portugal gets Portuguese rather than English. Unknown falls back to the
 * default, which is a normal thing for a store app to meet, not an error.
 */
export function normalizeLanguage(tag?: string | null): string {
  if (!tag) return DEFAULT_LANGUAGE;
  const t = tag.trim();
  if (BY_CODE.has(t)) return t;

  const lower = t.toLowerCase();
  for (const code of BY_CODE.keys()) {
    if (code.toLowerCase() === lower) return code;
  }
  const base = lower.split(/[-_]/)[0];
  if (BY_CODE.has(base)) return base;
  for (const code of BY_CODE.keys()) {
    if (code.split("-")[0].toLowerCase() === base) return code;
  }
  return DEFAULT_LANGUAGE;
}

const STORAGE_KEY = "app_language";

/**
 * The language to show right now.
 *
 * Order matters:
 *   1. What the user explicitly chose. Their choice outranks everything and
 *      is never silently overridden by a device setting.
 *   2. The device language. An Israeli phone set to Hebrew opens in Hebrew;
 *      the same phone set to English opens in English.
 *   3. English.
 *
 * The signed-in account's stored preference is applied separately, once the
 * profile loads — it cannot be consulted here because this runs before the
 * first request completes.
 */
export function detectLanguage(): string {
  try {
    const chosen = localStorage.getItem(STORAGE_KEY);
    if (chosen && BY_CODE.has(chosen)) return chosen;
  } catch {
    /* private mode or blocked storage — fall through to the device */
  }
  const device =
    typeof navigator !== "undefined"
      ? navigator.languages?.[0] || navigator.language
      : null;
  return normalizeLanguage(device);
}

export function rememberLanguage(code: string): void {
  try {
    localStorage.setItem(STORAGE_KEY, code);
  } catch {
    /* best effort: the account preference is the durable copy */
  }
}

/**
 * Point the document at a language and its writing direction.
 *
 * dir has to move with the language or an Arabic reader gets Arabic text in
 * a left-to-right layout, which is harder to read than English would have
 * been.
 */
export function applyDocumentLanguage(code: string): void {
  if (typeof document === "undefined") return;
  document.documentElement.lang = code;
  document.documentElement.dir = isRTL(code) ? "rtl" : "ltr";
}
